### DISCUSSION TITLE
Intellisense not recognising CMake  compiler defines

### DISCUSSION THREAD

**Discussion by daldr-ntml** [microsoft/vscode-cmake-tools#2542] [Q&A] [answered]
Hi, I'm using CMake Presets with vscode-cmake-tools. My selected preset applies compiler defines using target_compile_definitions() but VS Code Intellisense is not recognising those defines.  

Do I have to specify the defines explicitly in c_cpp_properties.json or should vscode-cmake-tools detect and apply them automatically?

(I am running on Windows but using an ARM cross-compiler).

**Comment by elahehrashedi** [comment, marked as answer]
vscode-cmake-tools should detect and apply defines automatically. But it looks like for your source file it is not being applied. 
I recommend you **open the source file** that is not configured properly and Run `CMake: log Diagnostics` on that and share the log with us.

**Reply by daldr-ntml** [reply]
@elahehrashedi Thanks for your answer. Here is the log you requested:

```
{
  "os": "win32",
  "vscodeVersion": "1.67.1",
  "cmtVersion": "1.10.5",
  "configurations": [
    {
      "folder": "c:\\SVNProj\\<snip>\\software\\elysion_arm",
      "cmakeVersion": "3.22.2",
      "configured": true,
      "generator": "Unix Makefiles",
      "usesPresets": true,
      "compilers": {}
    }
  ],
  "cpptoolsIntegration": {
    "isReady": true,
    "hasCodeModel": true,
    "activeBuildType": "Release",
    "buildTypesSeen": [
      "Release"
    ],
    "requests": [],
    "responses": [],
    "partialMatches": [],
    "targetCount": 1,
    "executablesCount": 1,
    "librariesCount": 0,
    "targets": [
      {
        "name": "elysion_arm",
        "type": "EXECUTABLE"
      }
    ]
  },
  "settings": [
    {
      "communicationMode": "automatic",
      "useCMakePresets": "auto",
      "configureOnOpen": true
    }
  ]
}
```

**Reply by bobbrow** [reply]
Cpptools hasn't requested any IntelliSense configuration information from us.  Two questions:
* Did you have the file open in the editor when you ran the Log Diagnostics command?
* Did you set CMake Tools as the configuration provider for cpptools?  (if not, you can use the `C/C++: Change Configuration Provider` command to set it to CMake Tools)

**Reply by daldr-ntml** [reply]
@bobbrow Thanks for your help. I have now set CMake Tools as the configuration provider for cpptools and IntelliSense is working as expected, i.e. it is disabling/enabling code according to my compiler defines.

The Output window shows:

`[13/05/2022, 08:31:19] Unable to resolve configuration with compilerPath "c:/program files/arm/development studio 2021.1/sw/armcompiler6.16/bin/armclang.exe".  Using "cl.exe" instead.`

though armclang.exe is present at that path. But Intellisense seems ok.

Thanks again.

**Reply by bobbrow** [reply]
Awesome!  Glad it's working now.  Cpptools isn't currently able to query the Keil compiler (related: https://github.com/microsoft/vscode-cpptools/issues/3222), so it picks a compiler that it knows so that you can get \*some\* IntelliSense for the standard headers.  If you need something more GCC-like, I believe you can override the `compilerPath` in `c_cpp_properties.json` and it should pick that alternate compiler instead.

**Reply by daldr-ntml** [reply]
@bobbrow Thanks for that link. These VS Code extensions are so useful and powerful, but it's difficult to remember all the details of how to configure them!
